File: solutions/javascript/robot-name/1/robot-name.js
const usedNames = new Set();

const generateRandomName = () => {
  const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  const digits = '0123456789';
  
  let name = '';
  for (let i = 0; i < 2; i++) {
    name += letters[Math.floor(Math.random() * letters.length)];
  }
  for (let i = 0; i < 3; i++) {
    name += digits[Math.floor(Math.random() * digits.length)];
  }
  return name;
};

export class Robot {
  #name;

  constructor() {
    this.reset();
  }

  get name() {
    return this.#name;
  }

  reset() {
    let newName;
    do {
      newName = generateRandomName();
    } while (usedNames.has(newName));

    usedNames.add(newName);
    this.#name = newName;
  }
}

Robot.releaseNames = () => {
  usedNames.clear();
};

File: solutions/kotlin/robot-name/1/src/main/kotlin/RobotName.kt
class Robot {
    private var _name: String? = null

    val name: String
        get() {
            if (_name == null) {
                _name = generateUniqueName()
            }
            return _name!!
        }

    fun reset() {
        _name = null
    }

    companion object {
        private val assignedNames = mutableSetOf<String>()

        private fun generateUniqueName(): String {
            while (true) {
                val letters = (('A'..'Z').random().toString() + ('A'..'Z').random())
                val digits = String.format("%03d", (0..999).random())
                val candidate = letters + digits
                if (assignedNames.add(candidate)) {
                    return candidate
                }
            }
        }
    }
}
